### PR TITLE
Feature/701: Finish recrypt

### PR DESCRIPTION
**Pull request by lkleisa**

### CODE REVIEW

File: spec/fabricators/encryptables/credentials_fabricator.rb [hunk 1]
```diff
@@ -5,12 +5,43 @@
...
+  cleartext_token { Faker::Internet.uuid }
+  cleartext_pin { Faker::Internet.ip_v4_address }
+  cleartext_email { Faker::Internet.email }
+  cleartext_custom_attr_label { Faker::Internet.user_name }
+  cleartext_custom_attr { Faker::Internet.password }
+  before_create do |encryptable, attrs|
+    encryptable.encrypt(attrs[:team_password])
+  end
+end
+
```
Comment (mtnstar) [inline spec/fabricators/encryptables/credentials_fabricator.rb:22]: create just one fabricator for credentials ....
add param to define which attrs should be used.
if no param, just create username, password

File: spec/controllers/api/encryptables_controller_spec.rb [hunk 2]
```diff
@@ -776,10 +775,38 @@
...
+
+      login_as(:alice)
+
+      expect(controller).not_to receive(:send_file)
+      expect_any_instance_of(CrudController).to receive(:render_entry).exactly(:once)
+
+      get :show, params: { id: encryptable_file.id,
+                           encryptable: ActionController::Parameters.new({
+                                                                           test: 1
+                                                                         }).permit! }, xhr: true
```
Comment (mtnstar) [inline spec/controllers/api/encryptables_controller_spec.rb:803]: permit? why is this used here? can't see any reason why we should permit attributes in a controller spec

File: spec/controllers/api/encryptables_controller_spec.rb [hunk 3]
```diff
@@ -776,10 +775,38 @@
...
+
+    it 'displays transferred encryptable and dont download it old encryption algorithm' do
+      login_as(:bob)
+
+      encryptable_file = prepare_transferred_encryptable(Crypto::Symmetric::Aes256)
+
+      login_as(:alice)
+
+      expect(controller).not_to receive(:send_file)
+      expect_any_instance_of(CrudController).to receive(:render_entry).exactly(:once)
```
Comment (mtnstar) [inline spec/controllers/api/encryptables_controller_spec.rb:798]: don't mock this ... try to find another way to get the data from the db

File: frontend/app/templates/teams/index.hbs [hunk 4]
```diff
@@ -1,14 +1,28 @@
...
-      <SearchResultComponent @q={{this.q}}/>
-    {{/if}}
-    {{#if @model}}
-      {{#each this.model as |team|}}
-        <Team::Show @team={{team}} />
-      {{/each}}
+    {{#if this.loading.isLoading}}
+        <div class="h-100 d-flex flex-column justify-content-center align-items-center">
+          <div class="lds-roller">
+            <div></div>
```
Comment (mtnstar) [inline frontend/app/templates/teams/index.hbs:6]: div div div ? isn't there a more elegant solution with css ?

File: spec/controllers/api/encryptables_controller_spec.rb [hunk 5]
```diff
@@ -776,10 +775,38 @@
       get :show, params: { id: encryptable_file.id }, xhr: true
     end
 
-    it 'displays transferred encryptable and dont download it' do
+    it 'download transferred encryptable new encryption algorithm' do
```
Comment (mtnstar) [inline spec/controllers/api/encryptables_controller_spec.rb:778]:
```suggestion
    it 'downloads transferred encryptable encrypted with most recent algorithm' do
```

File: spec/controllers/api/encryptables_controller_spec.rb [hunk 6]
```diff
@@ -776,10 +775,38 @@
...
+      encryptable_file = prepare_transferred_encryptable(Crypto::Symmetric::Aes256iv)
+
+      login_as(:alice)
+
+      expect(controller).to receive(:send_file).exactly(:once)
+
+      get :show, params: { id: encryptable_file.id }, xhr: true
+    end
+
+    it 'displays transferred encryptable and dont download it old encryption algorithm' do
```
Comment (mtnstar) [inline spec/controllers/api/encryptables_controller_spec.rb:790]: please make this text a bit clearer.

you could also group these inside a describe block

File: spec/controllers/api/teams_controller_spec.rb [hunk 7]
```diff
@@ -381,6 +414,47 @@
...
+          .and_return(nil)
+
+        default_folder = Folder.new(name: 'default', team: personal_team_bob)
+
+        personal_team_bob.folders = [default_folder, bob.inbox_folder]
+
+        get :index, params: { team_id: personal_team_bob.id }, xhr: true
+      end
+
+      it 'Receive method gets not called if encryptables are not transferred' do
```
Comment (mtnstar) [inline spec/controllers/api/teams_controller_spec.rb:436]: try not to describe test cases in a technical way. 

what does it or does it not in a user perspective?

**Review by mtnstar — CHANGES_REQUESTED** [review]
another huge PR we're going to merge hopefully soon!

some comments to check.

especially the test cases for the transferred enryptables vs. recryption need some improvement. 

will have another look on the PR for your next review request

File: spec/controllers/api/teams_controller_spec.rb [hunk 8]
```diff
@@ -574,6 +648,28 @@
 
   private
 
+  def prepare_transferred_encryptable
```
Comment (mtnstar) [inline spec/controllers/api/teams_controller_spec.rb:651]: maybe use a fabricator?

File: app/models/encryptable.rb [hunk 9]
```diff
@@ -102,6 +107,30 @@ def decrypt_attr(attr, team_password)
...
+  def encrypted_value_hash(attr)
+    data_attr = encrypted_data[attr]
+
+    data = data_attr.try(:[], :data)
+    iv = data_attr.try(:[], :iv)
+    { data: data, iv: iv }
+  end
+
+  def encrypt_with_encryption_class(cleartext_value, team_password, receiver_algorithm)
+    if cleartext_value.presence
```
[the remark this answers is not in the document]
Comment (lkleisa) [reply app/models/encryptable.rb:119]: Does this also handle the current line 126 with `{ data: nil, iv: nil }` ?

File: app/controllers/api/teams_controller.rb [hunk 10]
```diff
@@ -28,6 +28,35 @@ def index
 
   private
 
+  def fetch_team
+    @team = Team.find(team_id)
+    authorize @team, :team_member?
+
+    receive_transferred_encryptables(@team) if team.personal_team?
```
[the remark this answers is not in the document]
Comment (lkleisa) [reply app/controllers/api/teams_controller.rb:35]: No, you said that you would like to receive the transferred encryptables directly on click on the team. So it would now decrypt all transferred encryptable with transfer_password and then encrypt with personal_team_password on click on the personal_team. No matter if team is already recrypted or not

File: spec/controllers/api/encryptables_controller_spec.rb (hunk 2 above)
Comment (lkleisa) [reply spec/controllers/api/encryptables_controller_spec.rb:803]: I changed it, but it was already there from the old tests

File: spec/controllers/api/teams_controller_spec.rb (hunk 8 above)
Comment (lkleisa) [reply spec/controllers/api/teams_controller_spec.rb:675]: I moved the method to test helper, because we use it in encrpytables controller spec

File: spec/fabricators/encryptables/credentials_fabricator.rb (hunk 1 above)
Comment (lkleisa) [reply spec/fabricators/encryptables/credentials_fabricator.rb:19]: Robin said that it is too much work, so I made a [follow-up ticket](https://github.com/puzzle/cryptopus/issues/742)

**Comment by lkleisa** [comment]
The icon in encryptable/show with the lock needs a new design, otherwise feedback is implemented

File: spec/support/test_helper.rb [hunk 11]
```diff
@@ -13,6 +13,28 @@ def legacy_encrypt_private_key(private_key, password)
   encrypted_private_key
 end
 
+def prepare_transferred_encryptable(bob, alice, encryption_algorithm)
```
Comment (mtnstar) [inline spec/support/test_helper.rb:16]: can't we use a Fabricator for this?

so in the test we just could call: Fabricate(:transferred_encryptable, sender: bob, recipient: alice)

then use logic from  EncryptableTransfer for creating it ?

would find this way more clean than the helper here

File: spec/unit/serializers/team_list_serializer_spec.rb [hunk 12]
```diff
@@ -24,8 +24,9 @@
 
   context 'Some files transferred' do
```
Comment (mtnstar) [inline spec/unit/serializers/team_list_serializer_spec.rb:25]: same as with the other test class.

I know, this is not really related to the recrypt feature. but would be nice to clean it up anyway

**Review by mtnstar — CHANGES_REQUESTED** [review]
almost there !
check the feedback ... and let us discuss the changes first ... will be ready for merging soon, I'm sure !

File: spec/unit/utils/encryptable_move_handler_spec.rb [hunk 13]
```diff
@@ -15,7 +15,7 @@
     private_key = decrypt_private_key(bob)
     target_folder = folders(:folder2)
     team_password = target_folder.team.decrypt_team_password(bob, private_key)
-    Fabricate(:credential,
+    Fabricate(:credential_all_attrs,
```
[the remark this answers is not in the document]
Comment (lkleisa) [reply spec/unit/utils/encryptable_move_handler_spec.rb:18]: Robin said, create [follow up ticket](https://github.com/puzzle/cryptopus/issues/742)

File: spec/controllers/api/encryptables_controller_spec.rb [hunk 14]
```diff
@@ -764,32 +763,58 @@
...
     end
 
-    it 'download transferred encryptable' do
-      login_as(:bob)
+    context 'recrypt' do
+      it 'download transferred encryptable old encryption algorithm' do
+        login_as(:bob)
 
-      encryptable_file = prepare_transferred_encryptable
+        encryptable_file = prepare_transferred_encryptable(Crypto::Symmetric::Aes256)
```
Comment (mtnstar) [inline spec/controllers/api/encryptables_controller_spec.rb:770]: use fabricator !

File: spec/fabricators/encryptables/credentials_fabricator.rb [hunk 15]
```diff
@@ -5,12 +5,66 @@
 #  See the COPYING file at the top-level directory or at
 #  https://github.com/puzzle/cryptopus.
```
Comment (mtnstar) [inline spec/fabricators/encryptables/credentials_fabricator.rb:7]: move this fabricator to spec/fabricators/encrytable_fabricator.rb

**Review by mtnstar — CHANGES_REQUESTED** [review]
